src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,272 @@
+                                          size_t(0),
+                                          [](size_t result, const swarm_snode_map_t::value_type &pair) {
+                                            const ssize_t margin = pair.second.size() - EXCESS_BASE;
+                                            return result + std::max(margin, ssize_t(0));
+                                          });
+    LOG_PRINT_L2("Calculated excess: " << excess);
+    return excess;
+  };
+
+  /// Calculate threshold above which the excess should create a new swarm.
+  /// The threshold should be such that
+  /// 1. there is enough excess to create a new swarm of size NEW_SWARM_SIZE AND
+  /// 2. there is enough excess to leave IDEAL_SWARM_MARGIN excess in the existing swarms
+  prod_static size_t calc_threshold(const swarm_snode_map_t &swarm_to_snodes)
+  {
+    const size_t threshold = NEW_SWARM_SIZE + (swarm_to_snodes.size() * IDEAL_SWARM_MARGIN);
+    LOG_PRINT_L2("Calculated threshold: " << threshold);
+    return threshold;
+  };
+
+  prod_static const excess_pool_snode& pick_from_excess_pool(const std::vector<excess_pool_snode>& excess_pool, std::mt19937_64 &mt)
+  {
+    /// Select random snode
+    const auto idx = uniform_distribution_portable(mt, excess_pool.size());
+    return excess_pool.at(idx);
+  }
+
+  prod_static bool remove_excess_snode_from_swarm(const excess_pool_snode& excess_snode, swarm_snode_map_t &swarm_to_snodes)
+  {
+    auto &swarm_sn_vec = swarm_to_snodes.at(excess_snode.swarm_id);
+    const size_t original_size = swarm_sn_vec.size();
+    swarm_sn_vec.erase(std::remove(swarm_sn_vec.begin(), swarm_sn_vec.end(), excess_snode.public_key), swarm_sn_vec.end());
+    return original_size != swarm_sn_vec.size();
+  }
+
+  prod_static void get_excess_pool(size_t threshold, swarm_snode_map_t& swarm_to_snodes, std::vector<excess_pool_snode>& pool_snodes, size_t& excess)
+  {
+    /// Create a pool of all the service nodes belonging
+    /// to the swarms that have excess. That way we naturally
+    /// make the chances of picking a swarm proportionate to the
+    /// swarm size.
+    pool_snodes.clear();
+
+    if (threshold < MIN_SWARM_SIZE)
+      return;
+
+    excess = 0;
+    for (const auto &entry : swarm_to_snodes)
+    {
+      if (entry.second.size() > threshold)
+      {
+        excess += entry.second.size() - MIN_SWARM_SIZE;
+        for (const auto& sn_pk : entry.second)
+        {
+          pool_snodes.push_back({sn_pk, entry.first});
+        }
+      }
+    }
+  }
+
+  prod_static void create_new_swarm_from_excess(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    while (calc_excess(swarm_to_snodes) >= calc_threshold(swarm_to_snodes))
+    {
+      LOG_PRINT_L2("New swarm creation");
+      std::vector<crypto::public_key> new_swarm_snodes;
+
+      while (new_swarm_snodes.size() < NEW_SWARM_SIZE)
+      {
+        std::vector<excess_pool_snode> pool_snodes;
+        size_t excess;
```

Comment:
Not used. Do we need `get_excess_pool` to return excess? I is trivial to calculate given the size of pool_snodes.

Reply:
As per office discussion, you could either hoist the vector up (to justify the optimisation) or make `get_excess_pool ` return pool_snodes normally.

Reply:
As per discussion, not trivial...